source/tls_channel_handler.c

```diff
@@ -134,85 +135,135 @@ int aws_tls_ctx_options_init_client_mtls_from_path(
+    int custom_key_result;
 
-    /* certificate required, but there are multiple ways to pass it in */
-    if ((pkcs11_options->cert_file_path.ptr != NULL) && (pkcs11_options->cert_file_contents.ptr != NULL)) {
-        AWS_LOGF_ERROR(
-            AWS_LS_IO_TLS, "static: Both certificate filepath and contents are specified. Only one may be set.");
-        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+    if ((pkcs11_options->cert_file_contents.ptr != NULL) && (pkcs11_options->cert_file_path.ptr != NULL)) {
         goto error;
-    } else if (pkcs11_options->cert_file_path.ptr != NULL) {
+    } else if (pkcs11_options->cert_file_contents.ptr != NULL) {
+        custom_key_result = aws_tls_ctx_options_init_client_mtls_with_custom_key_operations(
+            options, allocator, pkcs11_handler, &pkcs11_options->cert_file_contents);
+    } else {
+        struct aws_byte_buf tmp_cert_buf;
         struct aws_string *tmp_string = aws_string_new_from_cursor(allocator, &pkcs11_options->cert_file_path);
-        int op = aws_byte_buf_init_from_file(&options->certificate, allocator, aws_string_c_str(tmp_string));
+        int op = aws_byte_buf_init_from_file(&tmp_cert_buf, allocator, aws_string_c_str(tmp_string));
         aws_string_destroy(tmp_string);
+
         if (op != AWS_OP_SUCCESS) {
+            aws_byte_buf_clean_up(&tmp_cert_buf);
             goto error;
         }
-    } else if (pkcs11_options->cert_file_contents.ptr != NULL) {
-        if (aws_byte_buf_init_copy_from_cursor(&options->certificate, allocator, pkcs11_options->cert_file_contents)) {
-            goto error;
-        }
-    } else {
-        AWS_LOGF_ERROR(AWS_LS_IO_TLS, "static: A certificate must be specified.");
-        aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
-        goto error;
-    }
 
-    if (aws_sanitize_pem(&options->certificate, allocator)) {
-        AWS_LOGF_ERROR(AWS_LS_IO_TLS, "static: Invalid certificate. File must contain PEM encoded data");
-        goto error;
+        struct aws_byte_cursor tmp_cursor = aws_byte_cursor_from_buf(&tmp_cert_buf);
+        custom_key_result = aws_tls_ctx_options_init_client_mtls_with_custom_key_operations(
+            options, allocator, pkcs11_handler, &tmp_cursor);
+        aws_byte_buf_clean_up(&tmp_cert_buf);
     }
 
-    /* Success! */
-    return AWS_OP_SUCCESS;
+    /**
+     * Calling aws_tls_ctx_options_init_client_mtls_with_custom_key_operations will have this options
+     * hold a reference to the custom key operations, but creating the TLS operations handler using
+     * aws_pkcs11_tls_op_handler_set_certificate_data adds a reference too, so we need to release
+     * this reference so the only thing (currently) holding a reference is the TLS options itself and
+     * not this function.
+     */
+    aws_custom_key_op_handler_release(pkcs11_handler);
+
+    return custom_key_result;
 
 error:
+
+    if (pkcs11_handler != NULL) {
+        aws_custom_key_op_handler_release(pkcs11_handler);
+    }
```

Comment:
same thing where I don't like it that 2 code paths in this file release the pkcs11_handler